File: packages/py/llmorch-cli/src/llmorch_cli/context.py
"""The Hexagonal seam for the Python CLI — twin of context.ts.

``build_context`` is the single place the CLI constructs an ``Orchestrator`` from
resolved config. The heavy ``llmorch_sdk`` import is deferred *inside* the function
so importing this module (or ``__main__``) stays cheap and ``llmorch --help`` never
loads the SDK. ``set_orchestrator_override`` is the test seam used by ``CliRunner``
to inject a ``FakeProvider``-backed orchestrator. The special ``fake`` provider
builds the in-``core`` ``FakeProvider`` for offline, key-free verification — it is
never registered globally, so ``providers`` still lists the four real providers.
"""

import json
import os
from collections.abc import Mapping
from dataclasses import dataclass, field
from typing import Any

_override: Any = None


def set_orchestrator_override(orch: Any) -> None:
    """Test seam: force ``build_context`` to return this orchestrator (or clear with None)."""
    global _override
    _override = orch


def _get_override() -> Any:
    return _override


@dataclass
class CliContext:
    orchestrator: Any
    flags: dict[str, Any] = field(default_factory=dict)


def _build_fake(env: Mapping[str, str]) -> Any:
    from llmorch_core.testing import FakeProvider

    kwargs: dict[str, Any] = {}
    text = env.get("LLMORCH_FAKE_TEXT")
    if text is not None:
        kwargs["text"] = text
    raw = env.get("LLMORCH_FAKE_STRUCTURED")
    if raw is not None:
        kwargs["structured"] = json.loads(raw)
    return FakeProvider(**kwargs)


def build_context(**flags: Any) -> CliContext:
    """Resolve config and construct the orchestrator (or return the test override)."""
    if _override is not None:
        return CliContext(orchestrator=_override, flags=flags)

    from .flags import resolve_flags

    resolved = resolve_flags(flags)
    if resolved.provider == "fake":
        return CliContext(orchestrator=_build_fake(os.environ), flags=flags)

    from llmorch_sdk import create_orchestrator  # deferred import (fast --help)

    config: dict[str, Any] = {"provider": resolved.provider}
    if resolved.model:
        config["default_model"] = resolved.model
    return CliContext(orchestrator=create_orchestrator(config), flags=flags)

File: packages/ts/cli/src/context.ts
/**
 * The Hexagonal seam for the CLI. `CommandContext` is what every command handler
 * receives; `buildOrchestrator` is the single place the CLI constructs an
 * `Orchestrator` from resolved config. Nothing here imports a provider SDK or a
 * concrete adapter package directly — the heavy `@llmorch/sdk` is pulled in
 * via a dynamic `import()` inside {@link buildOrchestrator} so it never loads for
 * `--help` or for a command (like `providers`) that makes no model call.
 */
import type { Orchestrator } from "@llmorch/core";

import type { ResolvedFlags } from "./flags.js";
import { defaultModelFor } from "./manifest.js";

/** Test/embedding seam injected into {@link run}. */
export interface RunDeps {
  /** Inject a prebuilt orchestrator (e.g. a `FakeProvider`) — skips the SDK build. */
  orchestrator?: Orchestrator;
  stdout?: (s: string) => void;
  stderr?: (s: string) => void;
  env?: NodeJS.ProcessEnv;
}

/** Everything a command handler is handed. Handlers parse/validate/render only. */
export interface CommandContext {
  /**
   * Resolve the shared orchestrator, built once and memoized. Commands that make
   * no model call (e.g. `providers`) never invoke this, so the SDK stays unloaded.
   */
  getOrchestrator: () => Promise<Orchestrator>;
  flags: Record<string, string | number | boolean | undefined>;
  args: string[];
  stdout: (s: string) => void;
  stderr: (s: string) => void;
  env: NodeJS.ProcessEnv;
}

/**
 * Construct an `Orchestrator` from resolved flags. The heavy `@llmorch/sdk`
 * (which transitively loads every provider SDK) is imported lazily here, so it
 * is only paid for when a command actually needs to talk to a model. The special
 * `fake` provider builds the in-`core` `FakeProvider` for offline, key-free
 * end-to-end verification — it is never registered globally, so `providers`
 * still lists exactly the four real providers.
 */
export async function buildOrchestrator(
  flags: ResolvedFlags,
  env: NodeJS.ProcessEnv,
): Promise<Orchestrator> {
  if (flags.provider === "fake") {
    const { FakeProvider } = await import("@llmorch/core");
    // Offline affordances so the binary can demo each modality with no live key:
    // LLMORCH_FAKE_TEXT seeds chat/stream output, LLMORCH_FAKE_STRUCTURED the
    // `structured` payload (must validate against the supplied schema).
    const text = env.LLMORCH_FAKE_TEXT;
    const structured = env.LLMORCH_FAKE_STRUCTURED
      ? JSON.parse(env.LLMORCH_FAKE_STRUCTURED)
      : undefined;
    return new FakeProvider({
      ...(text !== undefined ? { text } : {}),
      ...(structured !== undefined ? { structured } : {}),
    });
  }
  const { createOrchestrator } = await import("@llmorch/sdk");
  const config: Record<string, unknown> = { provider: flags.provider };
  const model = flags.model ?? defaultModelFor(flags.provider);
  if (model !== undefined) config.defaultModel = model;
  return createOrchestrator(config as never);
}
